File: apps/backend/app/schemas/webhook_routes.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field

from app.orders.types import OrderProduct, OrderType
from app.schemas.execution_intent import ProductMode
from app.schemas.ingestion_queue import QueueExecutionMode


class WebhookRouteOut(BaseModel):
    id: int
    user_id: int
    source: str
    name: str | None = None
    default_broker_key: str | None = None
    default_execution_mode: QueueExecutionMode
    default_product: str | None = None
    default_order_type: str | None = None
    policy: TradingViewRoutePolicy | None = None
    is_enabled: bool
    created_at: datetime
    updated_at: datetime


class TradingViewSizingMode(StrEnum):
    fixed_quantity = "fixed_quantity"
    fixed_amount = "fixed_amount"
    portfolio_percent = "portfolio_percent"


class PriceAndPctIn(BaseModel):
    price: float | None = Field(default=None, ge=0)
    pct: float | None = None


class TradingViewRoutePolicy(BaseModel):
    product_mode_default: ProductMode | None = None
    sizing_mode: TradingViewSizingMode | None = None
    fixed_quantity: int | None = Field(default=None, ge=1)
    fixed_amount: float | None = Field(default=None, ge=0)

    managed_exits_enabled: bool | None = None
    default_stop_loss: PriceAndPctIn | None = None
    default_target: PriceAndPctIn | None = None
    default_trailing_sl: PriceAndPctIn | None = None

    allow_payload_product: bool = True
    allow_payload_order_type: bool = True
    allow_payload_sizing: bool = True
    allow_payload_exits: bool = True


class WebhookRouteCreateRequest(BaseModel):
    name: str | None = Field(default=None, max_length=128)
    default_broker_key: str | None = Field(default=None, max_length=50)
    default_execution_mode: QueueExecutionMode = QueueExecutionMode.manual_review
    default_product: OrderProduct | None = None
    default_order_type: OrderType | None = None
    policy: TradingViewRoutePolicy | None = None


class WebhookRouteCreateResponse(BaseModel):
    route: WebhookRouteOut
    route_token: str


class WebhookRouteUpdateRequest(BaseModel):
    name: str | None = Field(default=None, max_length=128)
    default_broker_key: str | None = Field(default=None, max_length=50)
    default_execution_mode: QueueExecutionMode | None = None
    default_product: OrderProduct | None = None
    default_order_type: OrderType | None = None
    policy: TradingViewRoutePolicy | None = None
    is_enabled: bool | None = None

File: apps/frontend/src/lib/api/webhookRoutes.ts
import { apiRequest } from '@/lib/api/client'
import type { BrokerKey, OrderProduct, OrderType } from '@/lib/api/orders'

export type QueueExecutionMode = 'manual_review' | 'auto_dispatch'

export type ProductMode = 'delivery' | 'intraday' | 'carry_forward'

export type TradingViewSizingMode = 'fixed_quantity' | 'fixed_amount' | 'portfolio_percent'

export type PriceAndPct = { price?: number | null; pct?: number | null }

export type TradingViewRoutePolicy = {
  product_mode_default?: ProductMode | null
  sizing_mode?: TradingViewSizingMode | null
  fixed_quantity?: number | null
  fixed_amount?: number | null

  managed_exits_enabled?: boolean | null
  default_stop_loss?: PriceAndPct | null
  default_target?: PriceAndPct | null
  default_trailing_sl?: PriceAndPct | null

  allow_payload_product?: boolean
  allow_payload_order_type?: boolean
  allow_payload_sizing?: boolean
  allow_payload_exits?: boolean
}

export type WebhookRouteOut = {
  id: number
  user_id: number
  source: string
  name: string | null
  default_broker_key: BrokerKey | string | null
  default_execution_mode: QueueExecutionMode
  default_product: OrderProduct | string | null
  default_order_type: OrderType | string | null
  policy?: TradingViewRoutePolicy | null
  is_enabled: boolean
  created_at: string
  updated_at: string
}

export type WebhookRouteCreateRequest = {
  name?: string | null
  default_broker_key?: string | null
  default_execution_mode?: QueueExecutionMode
  default_product?: OrderProduct | null
  default_order_type?: OrderType | null
  policy?: TradingViewRoutePolicy | null
}

export type WebhookRouteCreateResponse = {
  route: WebhookRouteOut
  route_token: string
}

export type WebhookRouteUpdateRequest = {
  name?: string | null
  default_broker_key?: string | null
  default_execution_mode?: QueueExecutionMode | null
  default_product?: OrderProduct | null
  default_order_type?: OrderType | null
  policy?: TradingViewRoutePolicy | null
  is_enabled?: boolean | null
}

export async function listTradingViewRoutes(accessToken: string) {
  return apiRequest<WebhookRouteOut[]>('/api/v1/webhook-routes/tradingview', {
    method: 'GET',
    accessToken,
  })
}

export async function createTradingViewRoute(accessToken: string, payload: WebhookRouteCreateRequest) {
  return apiRequest<WebhookRouteCreateResponse>('/api/v1/webhook-routes/tradingview', {
    method: 'POST',
    accessToken,
    body: JSON.stringify(payload),
  })
}

export async function updateTradingViewRoute(accessToken: string, routeId: number, payload: WebhookRouteUpdateRequest) {
  return apiRequest<WebhookRouteOut>(`/api/v1/webhook-routes/tradingview/${routeId}`, {
    method: 'PATCH',
    accessToken,
    body: JSON.stringify(payload),
  })
}

